### ISSUE TITLE
System.OperationCanceledException when calling experimentResult.BestRun.Estimator.Fit

### ISSUE THREAD

**Issue by superichmann** [dotnet/machinelearning#6565]
**System Information (please complete the following information):**
 - OS & Version: Microsoft Windows 10 Pro
 - ML.NET Version: ML.NET 2.0.1
 - AutoML version:  AutoML 0.20.1
 - .NET Version: .NET 6.0

**Describe the bug**
As outlined in [this ](https://github.com/dotnet/machinelearning-samples/tree/main/samples/csharp/getting-started/AdvancedExperiment_AutoML#step-10-re-fit-best-pipeline)tutorial, a model can be refitted to include the latest data used in the validation phase of the AutoML experiment.

When calling this method, as in my case:
`        private static ITransformer RefitBestPipeline(ExperimentResult<RegressionMetrics> experimentResult,IDataView final)
        {
            RunDetail<RegressionMetrics> bestRun = experimentResult.BestRun;
            return bestRun.Estimator.Fit(final);
        }`
the return line causes the exception.

**To Reproduce**

1. Create a regression automl experiment in C#.
2. run the experiment on train validation data split
3. After experiment finished, call the `RefitBestPipeline` function with the best model and the entire IDataView (train+validation)
4. The error should occur on the return line

**Expected behavior**
The function should return the new ITransformer

**Screenshots, Code, Sample Projects**
```
System.OperationCanceledException
  HResult=0x8013153B
  Message=Operation was canceled.
  Source=Microsoft.ML.Core
  StackTrace:
   at Microsoft.ML.Runtime.Contracts.CheckAlive(IHostEnvironment env)
   at Microsoft.ML.Trainers.FastTree.DataConverter.MemImpl.InitializeBins(Int32 maxBins, IParallelTraining parallelTraining)
   at Microsoft.ML.Trainers.FastTree.DataConverter.MemImpl..ctor(RoleMappedData data, IHost host, Int32 maxBins, Single maxLabel, Boolean noFlocks, Int32 minDocsPerLeaf, PredictionKind kind, IParallelTraining parallelTraining, Int32[] categoricalFeatureIndices, Boolean categoricalSplit)
   at Microsoft.ML.Trainers.FastTree.DataConverter.Create(RoleMappedData data, IHost host, Int32 maxBins, Single maxLabel, Boolean diskTranspose, Boolean noFlocks, Int32 minDocsPerLeaf, PredictionKind kind, IParallelTraining parallelTraining, Int32[] categoricalFeatureIndices, Boolean categoricalSplit)
   at Microsoft.ML.Trainers.FastTree.ExamplesToFastTreeBins.FindBinsAndReturnDataset(RoleMappedData data, PredictionKind kind, IParallelTraining parallelTraining, Int32[] categoricalFeaturIndices, Boolean categoricalSplit)
   at Microsoft.ML.Trainers.FastTree.FastTreeTrainerBase`3.ConvertData(RoleMappedData trainData)
   at Microsoft.ML.Trainers.FastTree.FastForestRegressionTrainer.TrainModelCore(TrainContext context)
   at Microsoft.ML.Trainers.TrainerEstimatorBase`2.TrainTransformer(IDataView trainSet, IDataView validationSet, IPredictor initPredictor)
   at Microsoft.ML.Trainers.TrainerEstimatorBase`2.Fit(IDataView input)
   at Microsoft.ML.Data.EstimatorChain`1.Fit(IDataView input)
   at mlnet.AutoMLHelper.RefitBestPipeline(ExperimentResult`1 experimentResult, IDataView final)
```

**Additional context**
**Note:** I have read all of the other opened issues on this exception type and none of the proposed workarounds worked.
If relevant, In my case, the `mlContext` used to load the data is different then the one performing the experiment.
Perhaps the provided code example is too old now? (4 years) and irrelevant anymore?

**Comment by LittleLittleCloud** [comment]
You need to create a new MLContext for retraining purpose. As the original context has been canceled and can't be used to train a model.

**Comment by superichmann** [comment]
Hi again ;) I have read your comments in other issues but did not understand exactly how to do it.

The function is not called from `mlContext` but from `bestRun.Estimator.Fit`.

How do I enforce `bestRun.Estimator.Fit` to run from a new `mlContext`?

**Comment by LittleLittleCloud** [comment]
Ha,,, I see what you mean. It's possible that the refit experience is broken, I'll see what I can do here.

**Comment by jeffhandley** [comment]
I just ran into this bug using [2.0.1](https://www.nuget.org/packages/Microsoft.ML/2.0.1). @michaelgsharp / @ericstj, are we planning to publish a 2.0.2 with the fix from #6572?

**Comment by LittleLittleCloud** [comment]
The fix should already be released in 3.0.0-preview and after. Maybe try update to that version and see if you still hit this error?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by jeffhandley** [comment]
I did not encounter that same issue with the [3.0.0-preview.23266.6](https://www.nuget.org/packages/Microsoft.ML/3.0.0-preview.23266.6) package, but I encountered a different issue with model testing that I can't spend time troubleshooting right now. I'm thus blocked trying to adopt a version > 1.7.1. For a blocking issue on retraining, we should consider servicing 2.0.x with the fix rather than requiring users to adopt a new (preview) major version . For the time being, I'm rolling back to 1.7.1 to be unblocked.

My repro is when running the https://github.com/dotnet/issue-labeler (main branch) CreateMikLabelModel app with that repo set to use 2.0.x.

**Comment by ericstj** [comment]
>  I encountered a different issue with model testing that I can't spend time troubleshooting right now

Any chance you captured a call-stack of the problem using 3.0?  Might be something important to fix before GA.

**Comment by LittleLittleCloud** [comment]
> I did not encounter that same issue with the [3.0.0-preview.23266.6](https://www.nuget.org/packages/Microsoft.ML/3.0.0-preview.23266.6) package, but I encountered a different issue with model testing that I can't spend time troubleshooting right now. I'm thus blocked trying to adopt a version > 1.7.1. For a blocking issue on retraining, we should consider servicing 2.0.x with the fix rather than requiring users to adopt a new (preview) major version . For the time being, I'm rolling back to 1.7.1 to be unblocked.
> 
> My repro is when running the https://github.com/dotnet/issue-labeler (main branch) CreateMikLabelModel app with that repo set to use 2.0.x.

Hey @jeffhandley . I'm trying to reproduce the issue. Do I just need to clone that repo and `dotnet run`?

**Comment by jeffhandley** [comment]
@LittleLittleCloud No, that repo isn't very approachable for getting it running quickly to capture the repro.

But I reached the checkpoint on my work that I needed to hit, so I can circle back to it. Once I have that unrelated issue repro narrowed down, I'll create a new issue.

Back to the issue tracked here though, is there a workaround for the exception during `Estimator.Fit`, or is it a fully blocking issue in the 2.0.x versions?

**Comment by LittleLittleCloud** [comment]
@jeffhandley 
You can provide your own `ITrialRunner` which doesn't call `context.CancelExecution` when disposing

I would suggest simply copy and paste most of `SweepablePipelineRunner` from `src\Microsoft.ML.AutoML\AutoMLExperiment\Runner\SweepablePipelineRunner`, making change so it doesn't cancel execution at end, and register it with `AutoMLExperiment` using `SetTrialRunner<TTrialRunner>()`
